dao-contracts/src/voting/governance_voting.rs

```diff
@@ -0,0 +1,366 @@
+};
+
+use self::{
+    events::{
+        FormalVotingEnded, InformalVotingEnded, VoteCast, VotingContractCreated, VotingCreated,
+    },
+    voting::{Voting, VotingConfiguration, VotingResult, VotingType},
+};
+
+use casper_dao_utils::consts as dao_consts;
+
+use super::{vote::VotingId, Vote};
+
+/// The Governance Voting module.
+
+#[derive(Instance)]
+pub struct GovernanceVoting {
+    pub variable_repo: Variable<Option<Address>>,
+    pub reputation_token: Variable<Option<Address>>,
+    pub votings: Mapping<U256, Voting>,
+    pub votes: Mapping<(U256, Address), Vote>,
+    pub voters: Mapping<U256, Vec<Address>>,
+    pub votings_count: Variable<U256>,
+    pub dust_amount: Variable<U256>,
+}
+
+impl GovernanceVoting {
+    /// Initialize the module.
+    pub fn init(&mut self, variable_repo: Address, reputation_token: Address) {
+        self.variable_repo.set(Some(variable_repo));
+        self.reputation_token.set(Some(reputation_token));
+
+        emit(VotingContractCreated {
+            variable_repo,
+            reputation_token,
+            voter_contract: self_address(),
+        });
+    }
+
+    pub fn create_voting(
+        &mut self,
+        creator: Address,
+        stake: U256,
+        contract_to_call: Address,
+        entry_point: String,
+        runtime_args: RuntimeArgs,
+    ) {
+        let repo_caller = VariableRepositoryContractCaller::at(self.get_variable_repo_address());
+
+        let reputation_caller = ReputationContractCaller::at(self.get_reputation_token_address());
+
+        let informal_voting_time = repo_caller.get_variable(dao_consts::INFORMAL_VOTING_TIME);
+        let formal_voting_time = repo_caller.get_variable(dao_consts::FORMAL_VOTING_TIME);
+        let minimum_governance_reputation =
+            repo_caller.get_variable(dao_consts::MINIMUM_GOVERNANCE_REPUTATION);
+        let voting_id = self.votings_count.get();
+
+        let informal_voting_quorum = VariableRepositoryContractCaller::mul_by_ratio(
+            reputation_caller.total_onboarded(),
+            repo_caller.get_variable(dao_consts::INFORMAL_VOTING_QUORUM),
+        );
+        let formal_voting_quorum = VariableRepositoryContractCaller::mul_by_ratio(
+            reputation_caller.total_onboarded(),
+            repo_caller.get_variable(dao_consts::FORMAL_VOTING_QUORUM),
+        );
+
+        let voting_configuration = VotingConfiguration {
+            formal_voting_quorum,
+            formal_voting_time,
+            informal_voting_quorum,
+            informal_voting_time,
+            minimum_governance_reputation,
+            contract_to_call: Some(contract_to_call),
+            entry_point,
+            runtime_args,
+        };
+
+        let voting = Voting::new(voting_id, get_block_time(), voting_configuration);
+
+        // Add Voting
+        self.votings_count.set(voting_id + 1);
+        self.votings.set(&voting_id, voting);
+
+        emit(VotingCreated {
```

Comment:
When ending formal and informal votes you got two different events (`InformalVotingEnded` and `FormalVotingEnded`), but for starting the vote is just one. I think `InformalVotingCreated` and `FormalVotingCreated` should both exists, the difference would be a `FormalVotingCreated.informal_vote_id`.

Reply:
I would rather merge ending events into one than have additional type of event

Reply:
So please do.